src/main/java/com/example/petbuddybackend/service/user/CaretakerService.java

```diff
@@ -28,4 +37,54 @@ public Page<CaretakerDTO> getCaretakers(Pageable pageable, CaretakerSearchCriter
                 .findAll(spec, pageable)
                 .map(caretakerMapper::mapToCaretakerDTO);
     }
+
+    public Page<RatingDTO> getRatings(Pageable pageable, Long caretakerId) {
+        return ratingRepository.findAllByCaretakerId(caretakerId, pageable);
+    }
+
+    public void rateCaretaker(Long caretakerId, String clientUsername, int rating, String comment) {
+        assertCaretakerExists(caretakerId);
+        Long clientId = clientService.getClientIdByUsername(clientUsername);
+
+        if(clientId.equals(caretakerId)) {
+            throw new IllegalActionException("User cannot rate himself");
+        }
+
+        createOrUpdateRating(caretakerId, clientId, rating, comment);
+    }
+
+    public void deleteRating(Long caretakerId, String clientUsername) {
+        assertCaretakerExists(caretakerId);
+        Long clientId = clientService.getClientIdByUsername(clientUsername);
+        assertRatingExists(caretakerId, clientId);
+
+        ratingRepository.deleteById(new RatingKey(clientId, caretakerId));
+    }
+
+    private void assertRatingExists(Long caretakerId, Long clientId) {
+        if(!ratingRepository.existsById(new RatingKey(clientId, caretakerId))) {
+            throw new NotFoundException("Rating does not exist");
+        }
+    }
+
+    private void createOrUpdateRating(Long caretakerId, Long clientId, int rating, String comment) {
+        Rating ratingEntity = ratingRepository.findById(new RatingKey(clientId, caretakerId))
+                .orElse(
+                        Rating.builder()
+                                .clientId(clientId)
+                                .caretakerId(caretakerId)
+                                .build()
+                );
+
+        ratingEntity.setRating(rating);
+        ratingEntity.setComment(comment);
+
+        ratingRepository.save(ratingEntity);
+    }
+
+    private void assertCaretakerExists(Long caretakerId) {
```

Comment:
ja bym w tej metodzie od razu zwracal jesli istnieje a jesli nie to rzucal blad

Reply:
albo dodal druga, ktora to robi i w zaleznosci od potrzeby uzywac

Reply:
Troche refactor poszedł, ale dalej tylko sprawdzam czy coś istnieje. Jest to szybsze niż get całego entity

Reply:
slusznie